packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -1924,25 +1908,23 @@ export class HardhatNode extends EventEmitter {
       return this._runInPendingBlockContext(action);
     }
 
-    if (blockNumberOrPending === (await this.getLatestBlockNumber())) {
+    const latestBlockNumber = await this.getLatestBlockNumber();
+    if (blockNumberOrPending === latestBlockNumber) {
       return action();
     }
 
-    const block = await this.getBlockByNumber(blockNumberOrPending);
-    if (block === undefined) {
-      // TODO handle this better
-      throw new Error(
-        `Block with number ${blockNumberOrPending.toString()} doesn't exist. This should never happen.`
-      );
-    }
+    // const currentStateRoot = await this._context.vm().getStateRoot();
+    // const snapshotId = await this._context.vm().makeSnapshot();
+    await this._context.vm().setBlockContext(blockNumberOrPending);
 
-    const snapshot = await this._context.vm().makeSnapshot();
-    await this._setBlockContext(block);
     try {
       return await action();
     } finally {
-      await this._context.vm().restoreContext(snapshot);
-      await this._context.vm().removeSnapshot(snapshot);
+      await this._context.vm().restoreBlockContext(latestBlockNumber);
+      // await this._context.vm().restoreSnapshot(snapshotId);
```

Comment:
Can this line be removed?

Reply:
Resolved by https://github.com/NomicFoundation/hardhat/pull/4418/commits/9281dc1cb7410eb2ec79d38c66864033fd83542d